### PR TITLE
[SYCL] Diagnose LLVM memset that can't convert to SPIR-V

### PR DESCRIPTION
**Pull request by jtmott-intel**
There is no direct mapping of @llvm.memset.* intrinsic to SPIR-V. Special cases such as compile-time constant value and length have been made to work, but outside those limitations, there's no conversion and it currently fails in an assertion. This PR merely changes the assertion failure into a diagnostic message that -- hopefully -- will give the user a better idea of what the issue is.

I debated whether to put this diagnostic in SPIRVWriter.cpp like I did, vs somewhere else such as SemaSYCL.cpp. SemaSYCL has more source context, but the LLVM IR it generates is technically fine and correct, and it's only when and if we convert that LLVM IR to SPIR-V IR that we encounter an issue. Thoughts, opinions welcome. I looked to surrounding code for the correct way to emit a diagnositic in SPIRVWriter.cpp.

Before this PR:

    $ cat test.cpp

        #include <CL/sycl.hpp>

        SYCL_EXTERNAL
            void fn(int* ptr, int size) {
                ::std::memset(ptr, 0, sizeof(int) * size);
            }

    $ clang++ -fsycl test.cpp

        Stack dump:
        0.  Program arguments: /home/ubuntu/workspace/build/bin/llvm-spirv -o /tmp/test-7b6a5e-0a2800.spv -spirv-max-version=1.1 -spirv-debug-info-version=legacy -spirv-ext=+all,-SPV_INTEL_usm_storage_classes /tmp/test-d527b3.bc 
        1.  Running pass 'LLVMToSPIRV' on module '/tmp/test-d527b3.bc'.
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x5ab6dc)[0x55a02c0166dc]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x5a95a4)[0x55a02c0145a4]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x5a96f3)[0x55a02c0146f3]
        /lib/x86_64-linux-gnu/libpthread.so.0(+0x128a0)[0x7fbda5f638a0]
        /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7fbda4e18f47]
        /lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7fbda4e1a8b1]
        /lib/x86_64-linux-gnu/libc.so.6(+0x3042a)[0x7fbda4e0a42a]
        /lib/x86_64-linux-gnu/libc.so.6(+0x304a2)[0x7fbda4e0a4a2]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x1886b2)[0x55a02bbf36b2]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x188b27)[0x55a02bbf3b27]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x18b492)[0x55a02bbf6492]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x190966)[0x55a02bbfb966]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x193270)[0x55a02bbfe270]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x1933f3)[0x55a02bbfe3f3]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x4e0b86)[0x55a02bf4bb86]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0x1937e2)[0x55a02bbfe7e2]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0xda8da)[0x55a02bb458da]
        /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7fbda4dfbb97]
        /home/ubuntu/workspace/build/bin/llvm-spirv(+0xec03a)[0x55a02bb5703a]
        llvm-foreach: Aborted (core dumped)
        clang-12: error: llvm-spirv command failed with exit code 1 (use -v to see invocation)

After this PR:

    $ clang++ -fsycl test.cpp

        InvalidFunctionCall: Unexpected llvm intrinsic:
           call void @llvm.memset.p4i8.i64(i8 addrspace(4)* align 4 %0, i8 42, i64 %mul, i1 false)
        Translation of llvm.memset requires a const `length` argument
        llvm-foreach: 
        clang-12: error: llvm-spirv command failed with exit code 1 (use -v to see invocation)

Alexey Sachkov mentioned the separate PR https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/696 that adds LLVM memset conversion support to SPIR-V. That PR would make this one redundant. However, it's gone a month without activity and I can't predict how long before that PR gets merged, so in the meantime, still submitting this PR.

### CODE REVIEW

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -2244,7 +2244,10 @@ SPIRVValue *LLVMToSPIRV::transIntrinsicInst(IntrinsicInst *II,
     }
     Value *Len = MSI->getLength();
     if (!isa<ConstantInt>(Len)) {
-      assert(!"Can't translate llvm.memset with non-const `length` argument");
+      BM->getErrorLog().checkError(
+          isa<ConstantInt>(Len), SPIRVEC_InvalidFunctionCall,
+          toString(II) +
+              "\nCan't translate llvm.memset with non-const `length` argument");
```
Comment (premanandrao) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:2250]: I would change the diagnostic text to something like:
"\nTranslation of llvm.memset requires a non-const `length` argument"

to match other diagnostic formats used in this file.

**Review by premanandrao — COMMENTED** [review]
But otherwise, I am okay with a diagnostic being generated for this case, as well as your placement of it in SPIRVWriter.

When the limitation is lifted, the assertion would have needed removal, but with your change the diagnostic could be removed instead.

**Comment by bader** [comment]
What do you think about cherry-picking https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/696 to this repository?

**Comment by jtmott-intel** [comment]
@premanandrao Easy enough. Done.

@bader If it's done and working and ready, then that sounds like a great option. But I'm reluctant because it's a still-open PR with open review comments.

**Review by premanandrao — COMMENTED** [review]
Please fix the clang-format issue and add a test.
Otherwise, lgtm.

**Comment by jtmott-intel** [comment]
Test added, clang formatted.

GH says it needs an approving review, and after that I'm guessing it'll also block for the Jenkins failure. Does anyone know how to interpret the Jenkins results? I see a lot of `exit -1` and nothing else useful.

**Comment by jtmott-intel** [comment]
Jenkins issue fixed. The command output contained an offending path segment. I edited the PR summary and force pushed the branch to edit the git log.

**Comment by AlexeySotkin** [comment]
KhronosGroup/SPIRV-LLVM-Translator#696 is merged now and it should solve the problem. Still, I think the diagnostic emitted with this PR might be useful in case something will go wrong.

**Comment by premanandrao** [comment]
> [KhronosGroup/SPIRV-LLVM-Translator#696](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/696) is merged now and it should solve the problem. Still, I think the diagnostic emitted with this PR might be useful in case something will go wrong.

Currently @jtmott-intel's fix simply checks for a non-const length and issues the diagnostic for that.  Are you suggesting this check is still valid or something else?

**Comment by AlexeySotkin** [comment]
The check is still valid.

**Comment by jtmott-intel** [comment]
@AlexeySotkin If we want to still merge this change in, then GH still requires an approving review from a code owner.

**Comment by AlexeySotkin** [comment]
The code in this PR is unlikely to run due to KhronosGroup/SPIRV-LLVM-Translator#696. So I'm closing this PR.

**Pull request closed without merging**